### PR TITLE
transformations: (memref_stream) add streaming_region lowering

### PR DESCRIPTION
**Pull request by superlopuh**
Note stacked PR

Add a rewrite pattern from memref_stream.streaming_region to snitch_stream.streaming_region

### CODE REVIEW

File: xdsl/transforms/convert_memref_stream_to_snitch_stream.py [hunk 1]
```diff
@@ -70,6 +85,133 @@ def match_and_rewrite(
...
+            ),
+        ),
+    )
+
+
+def strides_for_affine_map(
+    affine_map: AffineMap, shape: Sequence[int], factor: int
+) -> list[int]:
+    """
+    Given an affine map mapping an iteration space to a domain, and a shape
```
Comment (nazavode) [inline xdsl/transforms/convert_memref_stream_to_snitch_stream.py:196]:
```suggestion
    Given an iteration space represented as an affine map (for indexing) and a shape (for bounds), returns the corresponding iteration strides for each dimension.
    
    The affine map must not have symbols.
```

**Review by nazavode — CHANGES_REQUESTED** [review]
LGTM, just completed a docstring that seemed an oversight.

File: xdsl/transforms/convert_memref_stream_to_snitch_stream.py [hunk 2]
```diff
@@ -70,6 +85,133 @@ def match_and_rewrite(
...
+def offset_map_from_shape(shape: Sequence[int], factor: int) -> AffineMap:
+    """
+    Given a list of lengths for each dimension of a memref, and the number of bytes per
+    element, returns the map from indices to an offset in bytes in memory. The resulting
+    map has one result expression.
+
+    e.g.:
+    ```
+    my_list = [1, 2, 3, 4, 5, 6]
+    shape = [2, 3]
```
Comment (compor) [inline xdsl/transforms/convert_memref_stream_to_snitch_stream.py:153]: is the shape `[2, 3]` or `[3, 2]`? Sorry, but the reversing somehow trips me up on what I expect.
Not sure what's actionable here.

**Review by compor — APPROVED** [review]

File: xdsl/transforms/convert_memref_stream_to_snitch_stream.py (hunk 2 above)
Comment (compor) [reply xdsl/transforms/convert_memref_stream_to_snitch_stream.py:153]: I agree, keeping it as it is in the other dialects (incl. upstream) would help.

Comment (superlopuh) [reply xdsl/transforms/convert_memref_stream_to_snitch_stream.py:153]: The good news is that as long as we get it wrong an even number of times then we'll get the right result in the end :)

**Review by nazavode — APPROVED** [review]

**Pull request merged**